src/ximea_ros_cam.cpp

```diff
@@ -869,9 +767,303 @@ void XimeaROSCam::triggerCb(const std_msgs::Empty::ConstPtr& msg) {
+            else
+                ROS_INFO_STREAM("Frame rate set to "
+                         << this->cam_framerate_set_ << "Hz.");  
+        } else { // default to free run
+            if(set( XI_PRM_ACQ_TIMING_MODE, XI_ACQ_TIMING_MODE_FREE_RUN))
+                ROS_ERROR_STREAM("Failed to set frame rate to free run.");
+            else  
+                ROS_INFO_STREAM("Frame rate set to free run.");  
+        }
+    }
+}
+
+void XimeaROSCam::sampleCameraTimestamp(void){
+    uint64_t nanoseconds;
+    ros::Time pre_stamp = ros::Time::now();
+    camera_timestamp_supported_ = get(XI_PRM_TIMESTAMP, nanoseconds) == XI_OK;
+    ros::Time post_stamp = ros::Time::now();
+
+    if(!camera_timestamp_supported_)        
+        return;       
+
+    const uint64_t billion = 1000000000; // c++11 has std::ratio, std::chrono and constexpr for this
+    timestamp_queue_.emplace_back(std::make_pair( pre_stamp, 
+            ros::Duration( nanoseconds/billion, nanoseconds%billion )));
+
+    // ROS_INFO_STREAM("Timestamps:" << // TODO delete or demote to DEGUG
+    //         " System: " << timestamp_queue_.back().first << 
+    //         " Camera: " << timestamp_queue_.back().second );
+
+    // monitor round trip speed
+    ros::Duration round_trip_duration(post_stamp - pre_stamp);
+    if(round_trip_duration > ros::Duration(0.0025)) // arbitary 2.5 ms (I see a 2 ms average with an xiC on my system)
+        ROS_WARN_STREAM_THROTTLE(3, "Camera timestamp query took " << round_trip_duration << " seconds." );    
+}
+
+ros::Time XimeaROSCam::iterpolateTimestamp(const XI_IMG& frame){
+    if(!camera_timestamp_supported_)  
+        return ros::Time::now();
+
+    sampleCameraTimestamp();
+
+    std::deque< std::pair<ros::Time,ros::Duration> >& queue = timestamp_queue_; // alias 
+    if(queue.size() < 2)
+    {  
+        ROS_ERROR("Not enough timestamp samples to interpolate.");
+        return ros::Time::now();
+    }
+
+    ros::Duration camera_stamp(frame.tsSec, frame.tsUSec * 1000);
+    // TODO adjust camera stamp to be the start of exposure (model dependant)
+    // consider using a precalculated bool to determin if exposure time should
+    // be deducted  
+    
+    bool clock_wrap = queue.front().second > queue.back().second; 
+    if(clock_wrap)
+    {
+        ROS_WARN("Clock wrap detected. Timestamp handling experimetal!");
+        // xiQ, xiD 
+        // 40 bit microsecond number - (overlaps after 305 hours)
+        // TimeStamp is recorded at the start of Data Readout 
+
+        // xiC, xiB, xiX, xiT 
+        // 64 bit 4-nanoseconds number (overlaps after 2339 years) (sic)
+        // TimeStamp is recorded at the start of Exposure.
+        // It is assumed that these cameras use the full tsSec value 
+
+        // untested experimental handling!
+        ros::Duration wrap_period;
+        ros::Duration rough_wrap_period = 
+                (queue.front().second - queue.back().second) + 
```

Comment:
I think its safe to use as is but I will review some more, in practice it is rare for the deque to get longer than 3.

Reply:
What do you think will be the case when deque becomes longer than 3, if any?

Reply:
Yes it is possible, it depends on the lag between image capture and retrieval via xiGetImage together with frame rate. All the same I would not expect it to get very big.

I guess it would be prudent to add a size check in the sample function that displays a warn when the deque goes over say 6, perhaps even pop the front if it blows up due to some unknown discrepancy between the XI_IMAGE stamp and the retrieved stamp.